### ISSUE TITLE
Flutter+Firestore iOS app crashes only when installed from iTunes

### ISSUE THREAD

**Issue by wkethman** [flutter/flutter#16361]
Does anyone have any thoughts?

The application can be compiled to run on both Android and iOS simulator (and through an attached iPhone in debug mode) - functions as expected. The application has been uploaded to Google Play and functions appropriately. The crash occurs after you have gone through the validation process, released by iOS (the first time it made it through and was crashing and the second time it was rejected because they caught it). The interesting part about this is that its difficult to debug because it crashes only after uploaded, not when debugging locally.

Dependencies:
```yaml
url_launcher: ^3.0.0
shared_preferences: ^0.4.0
cloud_firestore: ^0.4.0
```

There is a warning on iOS build that may be related to the crash but I would think this would cause a crash in development and not just when uploaded:

4.9.0 - [Firebase/Core][I-COR000003] The default Firebase app has not yet been configured. Add `[FIRApp configure];` (`FirebaseApp.configure()` in Swift) to your application initialization. 

## Logs

When uploaded to iTunes the application crashes on load with the following error:

```log
{"app_name":"Runner","timestamp":"2018-04-07 22:41:40.96 -0700","app_version":"1.1","slice_uuid":"","adam_id":0,"build_version":"2","bundleID":"com..rc","share_with_app_devs":false,"is_first_party":false,"bug_type":"109","os_version":"iPhone OS 11.2.5 (15D60)","incident_id":"","name":"Runner"}

Incident Identifier: C7E34EB2-0AA8-4D95-BBF2-4E323782A535
CrashReporter Key:   e7281b7b0d5ea708a1d85030a4736854635fe52e
Hardware Model:      iPhone9,3
Device Model:        D101AP
Process:             Runner [4283]
Path:                /private/var/containers/Bundle/Application/CEE81B18-1601-4704-8264-E6223725FE48/Runner.app/Runner
Identifier:          com..rc
Version:             2 (1.1)
Code Type:           ARM-64 (Native)
Role:                Foreground
Parent Process:      launchd.development [1]
Coalition:           com..rc [817]


Date/Time:           2018-04-07 22:41:40.8934 -0700
Launch Time:         2018-04-07 22:41:40.3694 -0700
OS Version:          iPhone OS 11.2.5 (15D60)
Baseband Version:    2.02.04_DEBUG
UDID:                d33d789e6dcbdf39cd323f95f8c5366344f52ba0
Report Version:      104

Exception Type:  EXC_CRASH (SIGABRT)
Exception Codes: 0x0000000000000000, 0x0000000000000000
Exception Note:  EXC_CORPSE_NOTIFY
Termination Description: DYLD, Symbol not found: _OBJC_CLASS_$_FIRApp | Referenced from: /private/var/containers/Bundle/Application/CEE81B18-1601-4704-8264-E6223725FE48/Runner.app/Frameworks/cloud_firestore.framework/cloud_firestore | Expected in: flat namespace | in /private/var/containers/Bundle/Application/CEE81B18-1601-4704-8264-E6223725FE48/Runner.app/Frameworks/cloud_firestore.framework/cloud_firestore
Triggered by Thread:  0
```

Filtered syslog:
```log
Timestamp              	Thread     Type        Activity             PID
2018-04-07 22:21:37.2989 -0700	0x868c 1024 0x0 102 hangtracerd: SpringBoard: Moving tailspin to spool: /var/root/Library/Caches/hangtracerd/spool/UIKit-runloop-SpringBoard-2018-04-07-222137.tailspin
2018-04-07 22:21:37.2991 -0700	0x868c 1024 0x0 102 hangtracerd: SpringBoard Log credits: 1 -> 0
2018-04-07 22:21:37.2991 -0700	0x868c 1024 0x0 102 hangtracerd: Daily Log credits: 18 -> 17
2018-04-07 22:25:48.7911 -0700	0x9d03 1024 0x0 102 hangtracerd: HTFenceTracking: Fence begin
2018-04-07 22:25:48.8725 -0700	0x9d03 1024 0x0 102 hangtracerd: HTFenceTracking: Fence end
2018-04-07 22:26:00.0015 -0700	0x9d03 1024 0x0 102 hangtracerd: XPC: Client started new connection: pid 3340
2018-04-07 22:26:00.0023 -0700	0x9d03 1024 0x0 102 hangtracerd: New proc: /private/var/containers/Bundle/Application/AC3B2C52-942C-44FD-A7E1-AEDA81916A25/Halos.app/Halos(3340)
2018-04-07 22:26:00.0023 -0700	0x9d03 1024 0x0 102 hangtracerd: Initialization complete now watching hangs for Halos(3340) | UIKit-runloop
2018-04-07 22:26:42.9680 -0700	0x9d03 1024 0x0 102 hangtracerd: HTFenceTracking: Fence begin
2018-04-07 22:26:43.0596 -0700	0x9d03 1024 0x0 102 hangtracerd: HTFenceTracking: Fence end
2018-04-07 22:27:02.8876 -0700	0x9d03 1024 0x0 102 hangtracerd: HTFenceTracking: Fence begin
2018-04-07 22:27:02.9051 -0700	0x9d03 1024 0x0 102 hangtracerd: HTFenceTracking: Fence end
2018-04-07 22:27:11.0515 -0700	0x9d03 1024 0x0 102 hangtracerd: HTFenceTracking: Fence begin
2018-04-07 22:27:11.0719 -0700	0x9d03 1024 0x0 102 hangtracerd: HTFenceTracking: Fence end
2018-04-07 22:28:17.4558 -0700	0x9d03 1024 0x0 102 hangtracerd: SpringBoard credits increased to: 1
2018-04-07 22:28:24.0847 -0700	0x9d03 1024 0x0 102 hangtracerd: Client disconnected: 3340
2018-04-07 22:30:06.9670 -0700	0x9d03 1024 0x0 102 hangtracerd: SpringBoard: Hang detected 0.61s (UIKit-runloop)
2018-04-07 22:30:06.9671 -0700	0x9d03 1024 0x0 102 hangtracerd: SpringBoard: Attempting to save tailspin at /var/root/Library/Caches/hangtracerd/tmp/UIKit-runloop-SpringBoard-2018-04-07-223006.tailspin
2018-04-07 22:30:07.0966 -0700	0x9d03 1024 0x0 102 hangtracerd: SpringBoard: Moving tailspin to spool: /var/root/Library/Caches/hangtracerd/spool/UIKit-runloop-SpringBoard-2018-04-07-223006.tailspin
2018-04-07 22:30:07.0969 -0700	0x9d03 1024 0x0 102 hangtracerd: SpringBoard Log credits: 1 -> 0
2018-04-07 22:30:07.0970 -0700	0x9d03 1024 0x0 102 hangtracerd: Daily Log credits: 17 -> 16
2018-04-07 22:33:43.3368 -0700	0x9d03 1024 0x0 102 hangtracerd: XPC: Client started new connection: pid 3821
2018-04-07 22:33:43.3369 -0700	0x9d03 1024 0x0 102 hangtracerd: New proc: /private/var/containers/Bundle/Application/D9839EC2-CF0D-4A1C-B36B-759650A629FF/All of Us.app/All of Us(3821)
2018-04-07 22:33:43.3371 -0700	0x9d03 1024 0x0 102 hangtracerd: Initialization complete now watching hangs for All of Us(3821) | UIKit-runloop
2018-04-07 22:33:54.6262 -0700	0x9d03 1024 0x2661f 102 hangtracerd: HTFenceTracking: Fence begin
2018-04-07 22:33:54.7030 -0700	0x9d03 1024 0x2661f 102 hangtracerd: HTFenceTracking: Fence end
2018-04-07 22:36:13.0167 -0700	0xdaa 1024 0x0 102 hangtracerd: Client disconnected: 3821
2018-04-07 22:37:27.8166 -0700	0xdaa 1024 0x0 102 hangtracerd: SpringBoard: Hang detected 0.73s (UIKit-runloop)

Thread 0 Crashed:
0   dyld                          	0x00000001035daf48 __abort_with_payload + 8
1   dyld                          	0x00000001035da71c abort_with_payload_wrapper_internal + 100 (terminate_with_reason.c:71)
2   dyld                          	0x00000001035da748 abort_with_payload + 12 (terminate_with_reason.c:93)
3   dyld                          	0x00000001035b0498 dyld::halt(char const*) + 312 (dyld.cpp:3824)
4   dyld                          	0x00000001035b3714 dyld::_main(macho_header const*, unsigned long, int, char const**, char const**, char const**, unsigned long*) + 7376 (dyld.cpp:6202)
5   dyld                          	0x00000001035ad21c _dyld_start + 68

Thread 0 crashed with ARM Thread State (64-bit):
    x0: 0x0000000000000006   x1: 0x0000000000000004   x2: 0x000000016ce26350   x3: 0x00000000000000c8
    x4: 0x000000016ce25f50   x5: 0x0000000000000000   x6: 0x0000000000000000   x7: 0x0000000000000000
    x8: 0x0000000000000020   x9: 0x0000000000000009  x10: 0x77656d6172662e65  x11: 0x756f6c632f6b726f
   x12: 0x72662e65726f7473  x13: 0x2f6b726f77656d61  x14: 0x69665f64756f6c63  x15: 0x0065726f74736572
   x16: 0x0000000000000209  x17: 0x0000000000000080  x18: 0xfffffff017751260  x19: 0x0000000000000000
   x20: 0x000000016ce25f50  x21: 0x00000000000000c8  x22: 0x000000016ce26350  x23: 0x0000000000000004
   x24: 0x0000000000000006  x25: 0x0000000102fd8000  x26: 0x00000001035ee000  x27: 0x00000001035edf88
   x28: 0x000000016ce27950   fp: 0x000000016ce25f20   lr: 0x00000001035da71c
    sp: 0x000000016ce25ee0   pc: 0x00000001035daf48 cpsr: 0x00000000
```

## Flutter Doctor

```terminal
[✓] Flutter (Channel dev, v0.2.9, on Mac OS X 10.13.4 17E199, locale en-US)
    • Flutter version 0.2.9 at /Users/wkethman/flutter
    • Framework revision 487e6bc91e (5 days ago), 2018-04-03 10:59:12 -0700
    • Engine revision 04f3c6557c
    • Dart version 2.0.0-dev.43.0.flutter-52afcba357

[!] Android toolchain - develop for Android devices (Android SDK 26.0.2)
    • Android SDK at /Users/wkethman/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 26.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)
    ✗ Android license status unknown.

[✓] iOS toolchain - develop for iOS devices (Xcode 9.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.3, Build version 9E145
    • ios-deploy 1.9.2
    • CocoaPods version 1.4.0

[✓] Android Studio (version 3.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)

[✓] Connected devices (1 available)
    • iPhone 6 • 9A7CCBF4-AF4B-415E-BB7E-9E8FAE3FBD78 • ios • iOS 11.3 (simulator)
```

**Comment by bizz84** [comment]
Same issue here.

### Crash error from Console

When running a CI build on a device, we have observed the following crash on the Console:

```
Exception Type:  EXC_CRASH (SIGABRT)
Exception Codes: 0x0000000000000000, 0x0000000000000000
Exception Note:  EXC_CORPSE_NOTIFY
Termination Description: DYLD, Symbol not found: _OBJC_CLASS_$_FIRApp | Referenced from: /private/var/containers/Bundle/Application/C77FC4F0-9400-4F09-92A1-33A374706F2D/Runner.app/Frameworks/firebase_auth.framework/firebase_auth | Expected in: flat namespace | in /private/var/containers/Bundle/Application/C77FC4F0-9400-4F09-92A1-33A374706F2D/Runner.app/Frameworks/firebase_auth.framework/firebase_auth
Triggered by Thread:  0
```

### flutter doctor

```
[✓] Flutter (Channel beta, v0.2.8, on Mac OS X 10.13.4 17E199, locale en-US)
[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.io/setup/#android-setup for detailed instructions).
      If Android SDK has been installed to a custom location, set $ANDROID_HOME to that location.
[✓] iOS toolchain - develop for iOS devices (Xcode 9.3)
[✗] Android Studio (not installed)
[!] Connected devices
    ! No devices available
```

## Setup

* We ran our build with `flutter build ios --release --preview-dart-2`
* We have followed the instructions from [#12086](https://github.com/flutter/flutter/issues/12086#issuecomment-332868543) and ensured our pod installation points to `ios-release`

```
 DEPENDENCIES:
-  - Flutter (from `<my_flutter_path>/bin/cache/artifacts/engine/ios`)
+  - Flutter (from `<my_flutter_path>/bin/cache/artifacts/engine/ios-release`)
  ...
 EXTERNAL SOURCES:
   Flutter:
-    :path: <my_flutter_path>/bin/cache/artifacts/engine/ios
+    :path: <my_flutter_path>/bin/cache/artifacts/engine/ios-release
```

### CI runner code

```
#!/usr/bin/env bash

set -e	# exit on first failed command
set -x  # print all executed commands to the terminal

cd ..

git clone -b beta https://github.com/flutter/flutter.git
export PATH=`pwd`/flutter/bin:$PATH

brew install --HEAD libimobiledevice
brew install ideviceinstaller
brew install ios-deploy

pip install six

flutter doctor

#sudo gem update cocoapods

cd git

flutter packages get

cd ios
pod install
cd ..

flutter test --preview-dart-2
flutter build ios --release --preview-dart-2
```

**NOTE**: On a test branch we have removed Firebase entirely from the project, and the release build no longer crashes on device.

**Comment by eseidelGoogle** [comment]
This appears to be the cause:
`Termination Description: DYLD, Symbol not found: _OBJC_CLASS_$_FIRApp | Referenced from: /private/var/containers/Bundle/Application/CEE81B18-1601-4704-8264-E6223725FE48/Runner.app/Frameworks/cloud_firestore.framework/cloud_firestore | Expected in: flat namespace | in /private/var/containers/Bundle/Application/CEE81B18-1601-4704-8264-E6223725FE48/Runner.app/Frameworks/cloud_firestore.framework/cloud_firestore
`
@collinjackson do you have any thoughts?  This appears to be cocoapods/firebase plugin related.

**Comment by eseidelGoogle** [comment]
FYI @cbracken

**Comment by eseidelGoogle** [comment]
This sounds very similar to https://github.com/flutter/flutter/issues/16014.

**Comment by eseidelGoogle** [comment]
Also mentioned in https://github.com/flutter/flutter/issues/15031, but I'm not sure that's the same.

**Comment by mravn-google** [comment]
Using Cocoapods 1.5.0 and the patches of https://github.com/flutter/flutter/pull/16443
https://github.com/flutter/plugins/pull/483 hopefully resolves this issue. Until we land the latter patch and publish new Firebase plugins, one will have to depend on a manually patched GitHub clone of flutter/plugins or similar.

**Comment by mravn-google** [comment]
Duplicate of #15031

**Comment by bizz84** [comment]
In my project I still experience the linking issues.

In my `pubspec.yaml` file I have this (note that `firebase_core` is not included because it's imported automatically with the other firebase packages): 

```
dependencies:
  flutter:
    sdk: flutter
  google_sign_in: 3.0.1
  firebase_auth: 0.5.4                                            # new
  firebase_database: 0.4.3                                       # new
  firebase_storage: 0.2.2
``` 

However, if I inspect my `firebase_core.podspec` I can still see this:

```
  s.pod_target_xcconfig = {
   'FRAMEWORK_SEARCH_PATHS' => '$(inherited) $(PODS_ROOT)/**',
   'OTHER_LDFLAGS' => '$(inherited) -undefined dynamic_lookup'
  }
```

So I added `firebase_core: 0.2.2` to my `pubspec.yaml`.

Now if I run `flutter packages get`, I get the following:
```
Running "flutter packages get" in flutter_firebase_example...
Because flutter_firebase_example depends on firebase_database 0.4.3 which depends on firebase_core
  ^0.1.0, firebase_core ^0.1.0 is required.
So, because flutter_firebase_example depends on firebase_core 0.2.2, version solving failed.
pub get failed (1)
```

So I need `firebase_core 0.2.2` to get past the linking issue, but `firebase_database 0.4.3` requires `firebase_core ^0.1.0`.

@mravn-google How to fix this?

**Comment by bizz84** [comment]
@mravn-google I think I have worked this out. See PR here: https://github.com/flutter/plugins/pull/490

**Comment by mravn-google** [comment]
You can add a dependency_override clause to your pubspec while we figure out if some of the Firebase packages need an upgrade to their firebase_core dependency.

**Comment by kroikie** [comment]
@wkethman

This issue has been moved to https://github.com/FirebaseExtended/flutterfire/issues/861. Any further collaboration will be done there.
